_posts/2024-02-20-Navigating-Dependency-Chaos-with-Lockfiles.md

```diff
@@ -0,0 +1,348 @@
+      Database 2
+    Data x processing
+      Package X
+      Package Y
+    Integration
+      Solution A
+      Platform B
+
+```
+
+_Dependency awareness and opportunity can be grouped into concerns and documented as part of a literature review (seen here as a [mind map](https://en.wikipedia.org/wiki/Mind_map))._
+{:.center}
+
+
+It can be helpful to reconsider existing knowledge on a topic area through formal or informal [literature review](https://en.wikipedia.org/wiki/Literature_review) (understanding that code within software is a type of literature) when thinking about the scale of decisions mentioned above.
+Outlining existing work through a literature review can help with [second-order thinking revision](https://en.wikipedia.org/wiki/Revision_(writing)#Reflection_in_the_revision_process) where we might benefit from reflecting on dependency decision-making again after an initial (first-order) creative process.
+Each potential dependency discovered through this process can be organized using [separation of concerns (SoC)](https://en.wikipedia.org/wiki/Separation_of_concerns) under specific _concern_ labels, or a general set of information which affects related code.
+Include dependencies within your project which will helpfully limit the code produced (or SoC _sections_) thereby reducing the overall amount of _concerns_ the project must maintain.
+
+```mermaid!
+flowchart LR
+
+    subgraph contextB["Context A"]
+        direction TB
+        processed1["data"]
+        widget
+        metadata1["metadata"]
+    end
+
+    subgraph contextA["Context B"]
+        direction TB
+        processed2["data"]
+        gizmo
+        metadata2["metadata"]
+    end
+
+    processed1 <-.-> | looosely coupled\nshared context | processed2
+    widget ~~~ | distinct solutions\nwith different names | gizmo
+    metadata1 ~~~ | distinct solutions\nwith the same names | metadata2
+```
+
+_Bounded contexts along with shared or distinct components can be used to help limit the complexity of a project in helpful ways._
+{:.center}
+
+The concept of [bounded context](https://martinfowler.com/bliki/BoundedContext.html) from [domain-driven design](https://en.wikipedia.org/wiki/Domain-driven_design) can sometimes be used to help distinguish what is in or out of scope for a particular project as a way of reducing complexity.
+Bounded context can be used as a way to draw abstract lines around a certain [span of control](https://en.wikipedia.org/wiki/Span_of_control) in order to align available resources (like time and people) with the focus of the project.
+It also can help promote [loose coupling](https://en.wikipedia.org/wiki/Loose_coupling) of software components in order to enable flexible design over time.
+Without these considerations and the use of dependencies we might face "endless" software [feature creep](https://en.wikipedia.org/wiki/Feature_creep) by continually adding new bounded contexts that are outside of our span of control (or resources).
```

Comment:
You touch on it in this section, but it could be helpful to briefly describe an example.

I might "outsource" something that is: 1) particularly error-prone, with a lot of edge cases, 2) something that needs constant maintenance to stay on top of, i.e. a "moving target" 3) something that requires esoteric or specialty knowledge. A good example of something like this in my work would be a library that handles dates and timezones.

Reply:
This is great, thanks! I'll incorporate this in.

Reply:
Added to the earlier section on "Why dependencies".